docs/proposal/20210823-gpu-support.md

```diff
@@ -0,0 +1,197 @@
+
+#### Story 3 - PCI passthrough for single node Customer
+
+Alex is an Engineer at a retail organization that requires a single GPU node. They use one node with GPU attached and want to keep things simple. Alex can simply add this GPU-connected machine to the cluster and that should do the job. While selecting nodes, Alex can use appropriate labels to run his AI/ML workload on this particular node. PCI passthrough will provide direct GPU support. The challenge that Alex can face is that using passthrough Alex wouldn't be able to migrate nodes.
+
+### Requirements
+
+#### Functional Requirements
+
+- FR1: CAPV MUST support vGPUs.
+- FR2: CAPV MUST support GPUDirect.
+
+- FR3: CAPV MUST support PCI Passthrough.
+
+#### Non-Functional Requirements
+
+- NFR1: Unit tests MUST exist for all 3 supports
+- NFR2: e2e tests MUST exist for 3 supports
+
+### Implementation Details/Notes/Constraints
+
+#### Proposed Changes
+
+```go
+
+// VirtualMachineCloneSpec is information used to clone a virtual machine.
+type VirtualMachineCloneSpec struct {
+  // Vgpu is the Vgpu configuration for this machine's VM.
+ // +optional
+ Vgpu []VGPUSpec `json:"vgpu,omitempty"`
+ // Pci is the Pci configuration for this machine's VM.
+ // +optional
+ Pci []PCISpec `json:"pci,omitempty"`
+}
+
+//PCISpec defines virtual machine's PCI configuration
+type PCISpec struct {
+ // Devices is the list of pci devices used by the virtual machine.
+ PciDevices []PCIDeviceSpec `json:"pcidevices"`
+}
+
+//VGPUSpec defines virtual machine's VGPU configuration
+type VGPUSpec struct {
+ // Devices is the list of vgpu devices used by the virtual machine.
+ VgpuDevices []VGPUDeviceSpec `json:"vgpudevices"`
+}
+
+//VGPUDeviceSpec defines virtual machine's VGPU configuration
+type VGPUDeviceSpec struct {
+ // ProfileName is the Profile Name of a virtual machine's VGPU, in string.
+ // Defaults to the eponymous property value in the template from which the
+ // virtual machine is cloned.
+ // +optional
+ ProfileName string `json:"profileName,omitempty"`
+}
+
+//PCIDeviceSpec defines virtual machine's PCI configuration
+type PCIDeviceSpec struct {
+
+ // DeviceId is the device ID of a virtual machine's PCI, in integer.
+ // Defaults to the eponymous property value in the template from which the
+ // virtual machine is cloned.
+ // +optional
+ DeviceId int32 `json:"deviceId,omitempty"`
+
+ // VendorId is the vendor ID of a virtual machine's PCI, in integer.
+ // Defaults to the eponymous property value in the template from which the
+ // virtual machine is cloned.
+ // +optional
+ VendorId int32 `json:"vendorId,omitempty"`
+}
+
+```
+
+#### Controller Changes
+
+- If VGPU is present, VGPU's `ProfileName` would be fetched from `VirtualMachineCloneSpec` and device object (`createPCIPassThroughDevice` from govmomi library) is formed using this VGPU `ProfileName`. This device object is added to
+- the slice of devices and then used by `vspheremachine_controller` to schedule nodes.
```

Comment:
The VM placement is dictated by the failureDomain present on the VirtualMachineSpec (maybe placement on a particular host or group op hosts), so what happens if the GPUProfile is not available on that host.

Reply:
Need a broader discussion on this.

Reply:
If we are using the Vsphere failure domain to dictate where the sphere VM gets created. It needs to have access to GPU, otherwise cloning will fail. It is users' responsibility to make sure that whatever failure domain they are trying to put the VM on(GPU VM). It must have the access to the required physical device.